extensions/notebook/src/book/bookTrustManager.ts

```diff
@@ -0,0 +1,137 @@
+		let trustableBookPaths = this.getTrustableBookPaths();
+		let trustableBooks: BookTreeItem[] = allBooks
+			.filter(bookItem => trustableBookPaths.some(trustableBookPath => trustableBookPath === path.normalize(bookItem.book.root)));
+		let isNotebookTrusted = this.isNotebookPartOfBooks(notebookUri, trustableBooks);
+
+		return isNotebookTrusted;
+	}
+
+	isNotebookPartOfBooks(notebookUri: string, books: BookTreeItem[]) {
+		let isPartOfNotebook: boolean = false;
+		let trustedBook: BookTreeItem = books.find(book => notebookUri.startsWith(path.resolve(book.root)));
+
+		if (trustedBook) {
+			let fullBookBaseUriWithContent = path.join(trustedBook.root, 'content');
+			let requestingNotebookFormattedUri = notebookUri.substring(fullBookBaseUriWithContent.length).replace('.ipynb', '');
+			let notebookInTOC = trustedBook.tableOfContents.sections.find(jupyterSection => {
+				let normalizedJupyterSectionUrl = jupyterSection.url && path.normalize(jupyterSection.url);
+				return normalizedJupyterSectionUrl === requestingNotebookFormattedUri;
+			});
+			isPartOfNotebook = !!notebookInTOC;
+		}
+
+		return isPartOfNotebook;
+	}
+
+	getTrustableBookPaths() {
+		let trustablePaths: string[];
+		let bookPathsInConfig: string[] = this.getTrustedBookPathsInConfig();
+
+		if (this.hasWorkspaceFolders()) {
+			trustablePaths = bookPathsInConfig
+				.map(trustableBookPath => this.workspaceDetails.workspaceFolders
+					.map(workspaceFolder => path.join(workspaceFolder.uri.fsPath, trustableBookPath))
+					.reduce((accumulator, currentTrustableBookPaths) => accumulator.concat(currentTrustableBookPaths)));
+
+		} else {
+			trustablePaths = bookPathsInConfig;
+		}
+
+		return trustablePaths;
+	}
+	getAllBooks(): BookTreeItem[] {
+		return this.books
+			.map(book => book.bookItems) // select all the books
+			.reduce((accumulator, currentBookItemList) => accumulator.concat(currentBookItemList)); // flatten them to a single list
+	}
+
+	setBookAsTrusted(bookRootPath: string): boolean {
+		let rootPathToStore: string = path.normalize(bookRootPath);
+		let matchingWorkspaceFolder: vscode.WorkspaceFolder = this.workspaceDetails.workspaceFolders
+			.find(ws => rootPathToStore.startsWith(path.normalize(ws.uri.fsPath)));
+
+		// if notebook is stored in a workspace folder, then store only the relative directory
+		if (matchingWorkspaceFolder) {
+			rootPathToStore.replace(path.normalize(matchingWorkspaceFolder.uri.fsPath), '');
+		}
+
+		let existingBooks: string[] = this.getTrustedBookPathsInConfig();
+
+		// if no match found in the configuration, then add it
+		if (!existingBooks.find(notebookPath => path.normalize(notebookPath) === rootPathToStore)) {
+			existingBooks.push(rootPathToStore);
+
+			// update the configuration
+			this.setTrustedBookPathsInConfig(existingBooks);
+			return true;
+		}
+		return false;
+	}
+
+	getTrustedBookPathsInConfig(): string[] {
+		let trustedBookPaths: string[] = this.trustedLocalBooks;
```

Comment:
Seems like we have a couple of scenarios here:
- Book is part of a workspace. What happens if a user opens the book in ADS outside of the workspace? Should it be trusted? I'm asking because I don't know the right answer 😄 
- Book isn't part of a workspace. This is the simpler scenario because it should always be trusted.

Can we leverage the same mementos mechanism somehow that we use for single notebooks? If you're not familiar with it, take a look at notebookServiceImpl.

Reply:
1. Currently, the way that it works now is that a separate book will be trusted, regardless if it is part of the workspace or not.
2. So if a book is not part of a workspace it is automatically trusted? What if I copied someone's book from a share into my drive? What constitutes the safety of the notebook? Currently I do not make this destinction in this change. 

Memento comment below

Reply:
I'm not sure if @chlafreniere is suggesting to use the actual same trusted notebook cache - but if the suggestion is just to use mementos in general that's available to extensions through extensionContext.globalState. 

Example : https://github.com/microsoft/azuredatastudio/blob/master/extensions/cms/src/extension.ts#L20

Reply:
Ah yes, I found this out after posting this comment 😁. Thanks for the pointer it helped out!
Leveraged the Mementos 😄